controllers/nstemplateset/namespaces.go

```diff
@@ -275,3 +287,97 @@ func getNamespaceName(request reconcile.Request) (string, error) {
 	}
 	return namespace, nil
 }
+
+// isUpToDateAndProvisioned checks if the obj has the correct Template Reference Label.
+// If so, it processes the tier template to get the expected roles and rolebindings and then checks if they are actually present in the namespace.
+func (r *namespacesManager) isUpToDateAndProvisioned(ns *corev1.Namespace, tierTemplate *tierTemplate) (bool, error) {
+	if ns.GetLabels() != nil &&
+		ns.GetLabels()[toolchainv1alpha1.TierLabelKey] == tierTemplate.tierName &&
+		ns.GetLabels()[toolchainv1alpha1.TemplateRefLabelKey] == tierTemplate.templateRef {
+
+		newObjs, err := tierTemplate.process(r.Scheme, ns.GetLabels()[toolchainv1alpha1.OwnerLabelKey], template.RetainAllButNamespaces)
+		if err != nil {
+			return false, err
+		}
+		processedRoles := []runtimeclient.Object{}
+		processedRoleBindings := []runtimeclient.Object{}
+		roleList := rbac.RoleList{}
+		rolebindingList := rbac.RoleBindingList{}
+		if err = r.AllNamespacesClient.List(context.TODO(), &roleList, runtimeclient.InNamespace(ns.GetName())); err != nil {
+			return false, err
+		}
+
+		if err = r.AllNamespacesClient.List(context.TODO(), &rolebindingList, runtimeclient.InNamespace(ns.GetName())); err != nil {
+			return false, err
+		}
+
+		for _, obj := range newObjs {
+			switch obj.GetObjectKind().GroupVersionKind().Kind {
+			case "Role":
+				processedRoles = append(processedRoles, obj)
+			case "RoleBinding":
+				processedRoleBindings = append(processedRoleBindings, obj)
+			}
+		}
+		// get the owner name from namespace
+		owner, exists := ns.GetLabels()[toolchainv1alpha1.OwnerLabelKey]
+		if !exists {
+			return false, fmt.Errorf("namespace doesn't have owner label")
+		}
+		//Check the names of the roles and roleBindings as well
+		for _, role := range processedRoles {
+			if found, err := r.containsRole(roleList.Items, role, owner); !found {
+				return false, err
+			}
+		}
+
+		for _, rolebinding := range processedRoleBindings {
+			if found, err := r.containsRoleBindings(rolebindingList.Items, rolebinding, owner); !found {
+				return false, err
+			}
+		}
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (r *namespacesManager) containsRole(list []rbac.Role, obj runtimeclient.Object, owner string) (bool, error) {
+	for _, val := range list {
+		if obj.GetObjectKind().GroupVersionKind().Kind == "Role" && val.GetName() == obj.GetName() {
+			// add owner label if doesn't exist
+			if _, exists := val.GetLabels()[toolchainv1alpha1.OwnerLabelKey]; !exists {
+				if val.Labels == nil {
+					val.Labels = make(map[string]string)
+				}
+				val.Labels[toolchainv1alpha1.OwnerLabelKey] = owner
+				if err := r.Client.Update(context.TODO(), &val); err != nil {
+					return false, err
+				}
```

Comment:
@MatousJobanek this is for existing roles which do not have the Owner label. If the label isn't present then any changes to those roles and rolebindings wouldn't trigger reconcile. See conversation here - https://github.com/codeready-toolchain/member-operator/pull/294#discussion_r722402786 Hope it clarifies?

Reply:
yes, I understand that. 
But if you return false - as I proposed, then it should re-apply all objects from the namespace template with all needed labels. This means that it would also fix all already existing resources. We can make a call to discuss it if you want.

Reply:
@MatousJobanek - ensureInnerNamespaceResources processes tierTemplate again, that's why I thought it would be easier to update labels here. Also IIRC, this piece of code wouldn't be needed once all roles have labels, since this is only for migration. I'm fine with either, if you think letting ensureInnerNamespaceResources handle it is still better, I can make that change. 
ps: sorry saw this late, and now its probably too late for you to take a call. Maybe we can discuss it next week if needed?

Reply:
I do apologize for my late response. I would let ensureInnerNamespaceResources handle it - after all, it's the responsibility of this function to keep all objects in sync. 
In addition, it will be less code and simpler.

Reply:
@MatousJobanek - done in 3ba1c87